chef/cookbooks/monasca/recipes/common.rb

```diff
@@ -0,0 +1,110 @@
+#
+# Copyright 2017 Fujitsu LIMITED
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitation.
+
+platform = node[:monasca][:platform]
+
+monasca_pkgs = platform[:packages]
+monasca_project = platform[:keystone][:tenant]
+monasca_roles = platform[:keystone][:roles]
+
+keystone_settings = KeystoneHelper.keystone_settings(node, @cookbook_name)
+
+register_auth_hash = {
+  user: keystone_settings["admin_user"],
+  password: keystone_settings["admin_password"],
+  tenant: keystone_settings["admin_tenant"]
+}
+
+log "monasca:common packages are #{monasca_pkgs}"
+log "monasca:common project is #{monasca_project}"
+log "monasca:common roles are #{monasca_roles}"
+
+monasca_pkgs.each do |pkg|
+  package pkg
+end
+
+keystone_register "monasca:common wakeup keystone" do
+  protocol keystone_settings["protocol"]
+  insecure keystone_settings["insecure"]
+  host keystone_settings["internal_url_host"]
+  port keystone_settings["admin_port"]
+  auth register_auth_hash
+  action :wakeup
+end
+
+keystone_register "monasca:common create tenant #{monasca_project} for monasca" do
+  protocol keystone_settings["protocol"]
+  insecure keystone_settings["insecure"]
+  host keystone_settings["internal_url_host"]
+  port keystone_settings["admin_port"]
+  auth register_auth_hash
+  tenant_name monasca_project
+  action :add_tenant
+end
+
+monasca_roles.each do |role|
+  keystone_register "monasca:common register #{role} role in #{monasca_project} tenant" do
+    protocol keystone_settings["protocol"]
+    insecure keystone_settings["insecure"]
+    host keystone_settings["internal_url_host"]
+    port keystone_settings["admin_port"]
+    auth register_auth_hash
+    role_name role
+    action :add_role
+  end
+end
+
+### recipes agents specific settings
+### single credentials all metrics@logs agents
+agents_settings = []
+
+# once agent is ready, uncomment following lines
+# if node["roles"].include?("monasca-agent")
+#   agents_settings.push(node[:monasca][:agent][:keystone])
+# end
+if node["roles"].include?("monasca-log-agent")
+  agents_settings.push(node[:monasca][:log_agent][:keystone])
+end
+
+unless agents_settings.empty?
+  agents_settings.each do |as|
+
+    keystone_register "monasca:common #{as["service_user"]} in #{as["service_tenant"]} user" do
```

Comment:
There are two problems with this resource:

1) `as["service_user"]` and `as["service_tenant"]` are empty for some reason. You can see this in the [gist](https://gist.githubusercontent.com/kornicameister/b35ad7dac3192b5e785186cf08e70e19/raw/a651d3945d66f1d6eb5e322d999c67916505426f/stack_trace2.out) you sent me: compare the code in `common.rb` and what actually gets passed to the `keystone_register` (in the bottom portion of the gist). The `tenant_name` and `role_name` parameters do not show up. When checking the proposal they were empty strings. At this point, my best guess is that setting a default in `attributes/default.rb` only has an effect if the attribute in question is unset in the data bag (rather than being an empty string). There'd be two solutions to this: provide a sensible value in the data bag or omit these two from `template-monasca.json` and set them to `required: false` in `template-monasca.schema`. In this case the values from `attributes/default.rb` should end up being used.

2) I manually set the missing `service_user` and `service_tenant` attributes in the data bag. At this point I got another failure because the service tenant didn't exist. Inserting the following resource above this line should fix this:

```
keystone_register "monasca:common create agent tenant #{as["service_tenant"]} for monasca" do
  protocol keystone_settings["protocol"]
  insecure keystone_settings["insecure"]
  host keystone_settings["internal_url_host"]
  port keystone_settings["admin_port"]
  auth register_auth_hash
  tenant_name as["service_tenant"]
  action :add_tenant
end

```